lib/ansible/parsing/vault/__init__.py

```diff
@@ -219,22 +219,90 @@ class VaultEditor:
 
     def __init__(self, password):
         self.vault = VaultLib(password)
-
+    
+    def _shred_file_custom(self, tmp_path):
+        """"Destroy a file, when shred (core-utils) is not available
+        
+        Unix `shred' destroys files "so that they can be recovered only with great difficulty with 
+        specialised hardware, if at all". It is based on the method from the paper 
+        "Secure Deletion of Data from Magnetic and Solid-State Memory", 
+        Proceedings of the Sixth USENIX Security Symposium (San Jose, California, July 22-25, 1996).
+        
+        We do not go to that length to re-implement shred in Python; instead, overwriting with a block
+        of random data should suffice. 
+        
+        See https://github.com/ansible/ansible/pull/13700 .
+        """
+        
+        file_len = os.path.getsize(tmp_path)
+        
+        passes = 3
+        with open(tmp_path,  "wb") as fh:
+            for _ in range(passes):
+                fh.seek(0,  0)
+                # get a random chunk of data 
+                data = os.urandom(min(1024*1024*2, file_len)) 
+                bytes_todo = file_len
+                while bytes_todo > 0:
+                    chunk = data[:bytes_todo]
+                    fh.write(chunk)
+                    bytes_todo -= len(chunk)
+                    
+                assert(fh.tell() == file_len)
+                os.fsync(fh)
+        
```

Comment:
* Slicing creates a copy so it might be better to avoid doing that inside of the inner loop
* We can move some calculations outside the loop (although it probably won't make much difference with just 3 passes).
* Instead of assert, we probably want to write a unittest to check that this does the right thing.  I'm okay with leaving the assert in until we  write a unittest, though.

So it would look something like this:
``` python

chunk_len = min(1024 * 1024 * 2, file_len)
with open(tmp_path, "wb") as fh:
    for _ in range(passes):
        fh.seek(0, 0)
        # get a random chunk of data
        data = os.urandom(chunk_len)
        for _ in range(0, file_len // chunk_len):
            fh.write(data)
        fh.write(data[:file_len % chunk_len])

        assert(fh.tell() == file_len)  # Remove this assert once we have unittests to check its accuracy
        os.fsync(fh)
```

Hopefully that's right and I'm not off by one byte :-)